### ISSUE TITLE
Thread support for Python

### ISSUE THREAD

**Issue by NQNStudios** [HaxeFoundation/haxe#9707]
Python supports multithreading (in Global Interpreter Locked form via `threading`, or in true form via `multiprocessing`). So is there any plan to support the [Haxe Thread API](https://api.haxe.org/sys/thread/Thread.html) on the Python target?

**Comment by NQNStudios** [comment]
I'm looking through the source code for the Neko Thread implementation and the way other libraries are implemented for the Python target. With some guidance I think I might be able to take this on as my first contribution to the Haxe language (I've contributed to libraries and docs before, but not the language itself).

**Comment by Simn** [comment]
The first step is to copy `std/sys/thread` to `std/python/_std/sys/thread`. Then you can change the `Thread` abstract so something like `abstract Thread(NativeThread)`, and start implementing its methods and defining `NativeThread` accordingly.

**Comment by NQNStudios** [comment]
Is `target.threaded` defined automatically based on whether the implementation is present, or will I have to go make sure `target.threaded` is defined true for Python? I haven't been able to figure out how this happens (I did `grep -r "target.threaded" .` and `grep -r "threaded" .` but I don't know ocaml so I can't make sense of it yet. (Time to learn some Ocaml??)

**Comment by NQNStudios** [comment]
I did get Haxe building correctly from source, and figured out that I should be running `(cd tests/threads && haxe build.hxml --python test.py && python test.py)` to verify my work. The first bug I encountered running this is the one thrown by the `#if target.threaded` check.

**Comment by Simn** [comment]
Check out `common.ml` and add `pf_supports_threads = true;` for the `Python` case in `get_config`. I think that's the only OCaml change necessary here.

**Comment by NQNStudios** [comment]
The suggested OCaml change worked. I've copied the `std/sys/thread` folder over as suggested, and I'm going to keep running the tests and fixing whichever error comes first. This has led me to think I should implement `Lock` first:

```
$ (cd tests/threads && haxe build.hxml --python test.py && python test.py)
Traceback (most recent call last):
  File "test.py", line 5813, in <module>
    haxe_EntryPoint.sleepLock = sys_thread_Lock()
NameError: name 'sys_thread_Lock' is not defined
```

This gives rise to two high-level questions:

* Should I rely on the Python `threading`, or `multiprocessing` module for the underlying Lock class (and all other underlying classes for this API)? My instinct (and what I'm going to do unless told otherwise) is to use multiprocessing because its behavior will more closely match what a Haxe programmer would expect on any other target.
* Should I use Python's `Lock` or [RLock](https://docs.python.org/3/library/multiprocessing.html#multiprocessing.RLock)? It looks like RLock supports the behavior specified in the Haxe sys.thread.Thread docs, and `Lock` does not. So I'm going with `RLock`, but I don't understand all the subtle differences mentioned in the Python docs.

**Comment by RealyUniqueName** [comment]
The answers to these questions could be given only by python users. What modules would you use for a python project? Are any of them "built-in" with the default python installation? And Haxe-like behavior is a good point too.

**Comment by NQNStudios** [comment]
`multiprocessing` and `threading` are both built-in. `multiprocessing` allows threaded programs to use more than one CPU core, while `threading` does not because... Python is weird. I don't know why `threading` exists to be honest.

Looking at the `RLock` class I think it satisfies the nested wait()/release() requirements of the Haxe API behavior, but this part of the docs worries me:

> A recursive lock must be released by the process or thread that acquired it.

So implementing our own releaseCount/waitCount over the more basic `Lock` class might be the way to go. Looking through the existing haxe source, I think the C# sys.thread implementation does something like this.

**Comment by NQNStudios** [comment]
Also, RLock does not allow release() to be called unless acquire() has already been called once. This violates the Haxe API's specifications so I'm even more sure that `Lock` will be the class to use.

**Comment by jonasmalacofilho** [comment]
[`multiprocessing`](https://docs.python.org/3/library/multiprocessing.html), as the name suggests, doesn't not use threads, but separate processes.  While the API follows `threading`, which helps, they are still fundamentally different, most notably in that they don't share (all) memory.  Processes also aren't "lightweight", as threads are usually expected to be.[1]

Additionally, messaging (using `Queue` or `Pipe`) involves serialization/de-serialization (`pickle`-ing).  Sharing memory seems to require wrapping things in `Value` or `Array`.  Both of these requirements are prone to cause a lot of nasty "surprises" if used to transparently implement `sys.thread`.

Maybe it would be best to implement `sys.thread` on top of `threading`, and support `multiprocessing` separately; `threading` threads are still useful, even with the GIL, for things that may take too long or block but that aren't themselves computationally intensive (for example, any type of I/O).

_[1] The start method is problematic too: `fork()`, while the default on Linux, is not safe to use indiscriminately, as it clones things like file descriptors and locks.  `forkserver()` may be a good alternative, but it's not immediately clear where it's supported.  And `spawn()`, the default on Windows and macOS, is documented to be (extra?) slow..._

_P.S. Threads are never guaranteed to run simultaneously; while the GIL is a ~~annoying~~ surprising bottleneck, even without it multithreading isn't guaranteed to improve performance.  It's just that most commonly we don't expect sharing code (but not data)  to also cause locking..._

**Comment by NQNStudios** [comment]
I had forgotten all these finer details about multiprocessing. You're right
that `threading` is the better choice. It suits my use-case just as well as
multiprocessing.

**Comment by Simn** [comment]
In my experience, the best order of implementation is the following:

1. `Thread.create` 
2. `Mutex`
3. `Deque`
4. The rest of `Thread`
5. Everything else you ever wanted to implemented in your life
6. `Lock`

**Comment by Simn** [comment]
Also, you should treat `tests/threads` like something that was written by people who really don't like you. You can isolate the `Issue4878` test and get that working first as it requires only `Thread.create` and `Mutex` support (well, and the correct behavior of `Sys.sleep`).

But, don't even look at `TestThreads`.

**Comment by Simn** [comment]
Oh nice, GitHub now allows assigning the author of an issue to that issue. The pact has been made.

**Comment by nanjizal** [comment]
@RealyUniqueName  
In reference to your question..
"The answers to these questions could be given only by python users. What modules would you use for a python project? Are any of them "built-in" with the default python installation? And Haxe-like behavior is a good point too."

When Haxe python came out I researched graphics approaches, I noticed most python users use tkinker or tools based on it for graphics and it may even come with Python out of the box.

For instance a class like this might also be worth adding.
http://www.randallmorgan.me/2018/12/20/simple-graphics-in-python/

It is much better that Haxe python users can implement and use graphics with standard python tools ( even if we feel that they maybe overkill ) rather than have to rely on if a large haxe toolkit that covers other targets, already many times support for edge targets with more cross toolkits are variable for less mainstream targets so it's useful for a target to have small graphics start pack.

While you can argue that it's not appropriate being high level it provides easy access much like flash, swing, canvas etc... and lets users start using Haxe python more easily and allows rapid prototyping and easy entry point to python developers wanting to try haxe, and visuals sell, and it's makes Haxe more appealing for any education establishments.

**Comment by nanjizal** [comment]
sorry on grammar was in a rush my correct later.

**Comment by Simn** [comment]
Yo, let's stay on topic. This has nothing to to with graphic on python or whatever.

**Comment by NQNStudios** [comment]
I have the `Issue4878.hx` test case passing.

**Comment by NQNStudios** [comment]
Since Deque is next on the implementation list, I'm guessing I should isolate the `DequeBrackets.hx` test next. I see that it used Thread.sendMessage() though, so maybe I'll have to implement that too to make sure my Deques are correct.

**Comment by NQNStudios** [comment]
Seeing the way Java implements `Thread.current()` with a WeakMap of java.lang.Thread to NativeThread, I tried to do the equivalent for Python and it looks like WeakMap isn't implemented for Python. I created issue #9722 for this, but if there's a suitable workaround, it shouldn't be a blocking issue for this one.

**Comment by NQNStudios** [comment]
I switched it to ObjectMap after verifying that `dict`, the underlying python type, would support Thread objects as keys. It should work because Thread objects are hashable.

**Comment by NQNStudios** [comment]
I'm stumped on implementing Lock. I'm modeling it after the Java implementation.

I can't get the DequeBrackets test to pass -- it usually hangs and the test fail message prints after quite a long pause. I've put some trace statements in to see the order things are happening, and give info on the Lock's releasedCount.

I think the problem is that python's `Condition.notify()` might not be exactly the same as Java's `Object.notify()`? For some reason, the notify() call doesn't always unblock the next thread.

My current working tree is [here](https://github.com/NQNStudios/haxe/commit/49819997f32149f0cd142c57ae568ec6c4c33bda) if anyone has a chance to look at it before I take another stab sometime tomorrow or next week. To run the isolated tests, rebuild Haxe and run this: `(cd tests/threads && ./haxe build.hxml -python test.py && python test.py)`

**Comment by RealyUniqueName** [comment]
Looks like you never clean up `threads` map. That means references to finished threads will stick forever.

**Comment by jonasmalacofilho** [comment]
@NQNStudios

For `Lock` I think [`threading.Semaphore`](https://docs.python.org/3.8/library/threading.html#semaphore-objects) will simpify things: as far as I could check, it has essentially the same semantics, except from the initial released count.

I tried the following implementation and it seemed to work... the DequeBrackets tests still hang, but seemly for different reasons.

```haxe
@:forward(release)
abstract Lock(NativeSemaphore) {
	public inline function new() {
		this = new NativeSemaphore(0);
	}

	public inline function wait(?timeout:Float):Bool {
		return this.acquire(true, timeout);
	}
}
```

<details><summary>Externs for `Lock` implementation</summary>

```haxe
@:pythonImport("threading", "Semaphore")
@:native("Lock")
private extern class NativeSemaphore {
	function new(value:Int);
	function acquire(blocking:Bool = true, ?timeout:Float):Bool;
	function release():Void;
}
```
</details>

_(You could also make `Mutex` an abstract over `RLock`)_

<details><summary>Edited... silly me, `collections.deque()` is threadsafe!  Sorry about that!</summary>

~~The next apparent cause of errors was `Deque`.  I noticed you were using `collections.deque`, and that makes sense, since we append to both sides and no class in `queue` supports that.~~

~~But `collections.deque` isn't threadsafe, so a lock of some sort is necessary.  Since you had mentioned it, I looked into `threading.Condition`, and I think its `wait_for(<predicate>)` method is great here.~~

```haxe
using python.internal.UBuiltins;

class Deque<T> {
	var deque:NativeDeque<T>;
	var lock:NativeCondition;

	public function new() {
		deque = new NativeDeque<T>();
		lock = new NativeCondition();
	}

	public function add(i:T) {
		lock.acquire();
		deque.append(i);
		lock.notify();
		lock.release();
	}

	public function push(i:T) {
		lock.acquire();
		deque.appendleft(i);
		lock.notify();
		lock.release();
	}

	public function pop(block:Bool):Null<T> {
		var ret = null;
		lock.acquire();
		if (block) {
			lock.wait_for(() -> deque.bool());
			ret = deque.popleft();
		} else if (deque.bool()) {
			ret = deque.popleft();
		}
		lock.release();
		return ret;
	}
}

@:pythonImport("collections", "deque")
@:native("deque")
extern class NativeDeque<T> {
	function new();
	function append(x:T):Void;
	function appendleft(x:T):Void;
	function popleft():T;
}

@:pythonImport("threading", "Condition")
@:native("Condition")
private extern class NativeCondition {
	function new(?lock:Dynamic);
	function acquire(blocking:Bool = true, timeout:Float = -1):Bool;
	function release():Void;
	function wait(?timeout:Float):Bool;
	function wait_for(predicate:()->Bool, ?timeout:Float):Bool;
	function notify(n:Int = 1):Void;
	function notify_all():Void;
}
```
</details>

This seems to help, and `DequeBrackets` now seems to hang on an internal `threading._shutdown` method.  It's probably related to having to use `ObjectMap` instead of `WeakMap`.  I tried to find a workaround for that, but I got nowhere.

For reference, with the changes above:

```
Issue3767: ok
Issue4878: ok
Issue8063: ok
DequeBrackets: hangs on threading._shutdown
WeirdTreeSum: hangs on threading._shutdown
TestThreads: Error NameError("name 'sys_thread_Tls' is not defined") for para\meters [...]
```

**Comment by NQNStudios** [comment]
Ha, I was pretty dang sure I had read in the docs that `collections.deque` is threadsafe. Glad I didn't goof on that.

**Comment by NQNStudios** [comment]
This part of the Semaphore documentation confuses me:

> A semaphore manages an internal counter which is decremented by each acquire() call and incremented by each release() call. The counter can never go below zero; when acquire() finds that it is zero, it blocks, waiting until some other thread calls release().

But the Haxe docs for Lock say this:

> The order of the release and wait calls is irrelevant. That is, a Lock can be released before anyone waits for it. In that case, the wait call will execute immediately.

The Java implementation of Lock [seems to let its counter go below 0](https://github.com/HaxeFoundation/haxe/blob/39ce020135f8b687b08ad99370382a54f46f88a5/std/java/_std/sys/thread/Lock.hx#L39). Is that required to fulfill the Haxe Lock spec?

**Comment by jonasmalacofilho** [comment]
The Java implementation goes bellow zero, but then the waiter blocks (or returns `false`), which is the important thing.

The counter isn't part of Haxe's specification.  _(non-authoritative statement, but it isn't part of the documented contract)_

---

I done goofed with the atomicity of `collections.deque` and ended not checking your `Deque` before starting to mess with it :facepalm: 

Now I noticed a small problem in `Deque.pop`, it isn't atomic:

```haxe
public function pop(block:Bool):Null<T> {
    while (block && nativeDeque.len() == 0) { }
    return try {
        nativeDeque.popleft();  // other thread could have already popped the value
    } catch (indexError:Dynamic) {
        null;
    };
}
```

**Comment by NQNStudios** [comment]
Does this commit fix the bug in pop()? https://github.com/NQNStudios/haxe/commit/16bdd85746c78a0a2cd3e01bf63e2caf58ac4eb8

EDIT: I was using a Lock instead of a Mutex. This commit should be better 62fcf0f773ca33ca276dc708c623c3cf499acad9

**Comment by NQNStudios** [comment]
@RealyUniqueName 

> Looks like you never clean up threads map. That means references to finished threads will stick forever.

This is true. I'm looking at the Java implementation, though, and can't see anywhere that it cleans up its map, either.

The simplest way I can think of to automatically clean the map, would be to wrap the `callb` functions something like this:

```
public static function create(callb:Void->Void):Thread {
    var nt: NativeThread = null;
    var wrappedCallB = () -> { callb(); threads.remove(nt); }
    nt = new NativeThread(null, wrappedCallB);
    nt.start();
    var t = new Thread(nt);
    threads.set(nt, t);
    return t;
}
```

Pretty janky. But the alternative would be running a daemon thread or having a check somewhere that checks `isAlive()` on all the native threads in the map, right? Which also seems weird.

**Comment by jonasmalacofilho** [comment]
> Does this commit fix the bug in pop()? NQNStudios@16bdd85
> EDIT: I was using a Lock instead of a Mutex. This commit should be better 62fcf0f

That "works", but the `while (block && nativeDeque.len() == 0) { }` is busy waiting (=is a "hot" loop), no?

In the end, there might be a point in using a Python `Condition` after all.  I can't see another way to _efficiently_ block until there's something to pop from the native deque.

It would essentially look like the code I posted before (and later hid), since you still need to acquire the condition to `notify` on it, and our `add`/`push` methods need to notify threads blocked at `pop`.

---

> This is true. I'm looking at the Java implementation, though, and can't see anywhere that it cleans up its map, either.

The Java implementation stores the Thread objects in an WeakMap, so the GC will trigger the cleanup automatically when there are no other strong references.

A "manual finalizer" of sorts may work on non-main threads.  Your code would need a mutex on the `threads` variable, but seems like something that could work...

But I'm worried this wont work on the main thread, assuming `Thread.current()` has been called on it (and thus set an entry for it in the `threads` map).  : \

**Comment by NQNStudios** [comment]
I got all the tests to pass locally! I assume my code isn't the most elegant and might have some formatting issues, but I'll open the PR after I find a solution for cleaning the `threads` map.
